src/nccl_ofi_net.cpp

```diff
@@ -578,11 +578,39 @@ int nccl_net_ofi_info_properties(nccl_net_ofi_plugin_t *plugin, struct fi_info *
 
 	props->max_mr_key_size = nic_prov->domain_attr->mr_key_size;
 
-
 	props->dmabuf_support = ((nic_prov->caps & FI_HMEM) != 0) &&
 		FI_VERSION_GE(nic_prov->fabric_attr->api_version, FI_VERSION(1, 20)) &&
 		nccl_ofi_dmabuf_viable()
 		;
+	if (props->dmabuf_support && strncmp("efa", nic_prov->fabric_attr->prov_name, strlen("efa")) == 0) {
+		// Generations 1-3 of EFA have a firmware issue that can result
+		// in communication failures with MRs that cover a large number
+		// of page entries.  This is not usually a problem, because page
+		// merging greatly reduces the number of page entries in the MR.
+		// However, the RDMA subsystem in the Linux kernel did not
+		// properly execute page merging for dmabuf entries until a
+		// recent (Feb 2025) patch, and the lack of page merging
+		// increased the probability of hitting the EFA issue.  Testing
+		// for the fixed kernel version is effectively impossible (the
+		// issue can also be fixed in the EFA kmod itself, and backports
+		// are likely, so a simple kernel version check is
+		// insufficient), so instead we only support dmabuf by default
+		// in Generation 4 of EFA.  When the communication failure issue
+		// is resolved in previous generations, this code will be
+		// removed and dmabuf will be available by default everywhere.
+		if (nic_prov->nic == NULL || nic_prov->nic->device_attr == NULL) {
+			props->dmabuf_support = false;
```

Comment:
A trace here too for the TCP like cases where we know this is why it was disabled?

Reply:
FWIW, this was more of a bozo check.  The TCP case can't get here, because we're in an EFA-only block.  But added the trace.